helix-term/src/commands/typed.rs

```diff
@@ -1365,21 +1366,60 @@ fn lsp_stop(
 
     let doc = doc!(cx.editor);
 
-    let ls_id = doc
-        .language_server()
-        .map(|ls| ls.id())
-        .context("LSP not running for the current document")?;
+    // TODO this stops language servers which may be used in another doc/language type that uses the same language servers
+    // I'm not sure if this is really what we want
+    let ls_shutdown_names = doc
+        .language_servers()
+        .map(|ls| ls.name())
+        .collect::<Vec<_>>();
 
-    let config = doc
-        .language_config()
-        .context("LSP not defined for the current document")?;
-    cx.editor.language_servers.stop(config);
+    for ls_name in &ls_shutdown_names {
+        cx.editor.language_servers.stop(ls_name);
+    }
+
+    let doc_ids_active_clients: Vec<_> = cx
+        .editor
+        .documents()
+        .filter_map(|doc| {
+            let doc_active_ls_ids: Vec<_> = doc
+                .language_servers()
+                .filter(|ls| !ls_shutdown_names.contains(&ls.name()))
+                .map(|ls| ls.id())
+                .collect();
+
+            let active_clients: Vec<_> = cx
+                .editor
+                .language_servers
+                .iter_clients()
+                .filter(|client| doc_active_ls_ids.contains(&client.id()))
+                .map(Clone::clone)
+                .collect();
+
+            if active_clients.len() != doc.language_servers().count() {
+                Some((doc.id(), active_clients))
+            } else {
+                None
+            }
+        })
+        .collect();
+
+    for (doc_id, active_clients) in doc_ids_active_clients {
+        let doc = cx.editor.documents.get_mut(&doc_id).unwrap();
+
+        let stopped_clients: Vec<_> = doc
+            .language_servers()
+            .filter(|ls| {
+                !active_clients
+                    .iter()
+                    .any(|active_ls| active_ls.id() == ls.id())
+            })
```

Comment:
Yes, you're right, sometimes I'm caught up a little bit in FP Kool-Aid (even there, might be a better solution IMHO). Wasn't too happy with it, when I've written it either.

New version should be much shorter and more concise, thanks for the tip.
Haven't come around extra allocations for the language server names though, but I think it's definitely negligible, and I have focused on readability here instead of efficiency (since it's only executed at `lsp_stop` and the vecs are very small).